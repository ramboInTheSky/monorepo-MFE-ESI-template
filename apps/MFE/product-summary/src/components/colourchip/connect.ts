import {connect} from "react-redux"
import State from "../../models/state"
import {setSelectedColourwayAction, setAnimateFavouriteIcon} from "../../ducks/productSummary"

export const mergeProps = (state: State, store: any, ownProps) => ({
    ...state,
    ...ownProps,
    setMatchingColourWay: () => {
        store.dispatch(setSelectedColourwayAction(ownProps.itemNumber))
        store.dispatch(setAnimateFavouriteIcon(false))
    },
})

export default connect(null, null, mergeProps)
