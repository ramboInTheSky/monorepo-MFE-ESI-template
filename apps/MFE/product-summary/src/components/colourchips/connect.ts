import {connect} from "react-redux"
import {ColourwayData} from "../../models/searchApi"
import State from "../../models/state"
import {formatPdpLink, formatTitle} from "../../utils/colourwayBuilders"
import {selectPriceType} from "../../utils/priceFormatter/selectPriceType"
import {IMAGE_EXTENSION, LG_IMAGE_PATH} from "../../config/constants"

export const mapStateToProps = (state: State) => {
    const {
        productSummary: {
            summaryData: {colourways, imageCdnUrl, baseUrl, fit, title, department, productImageUrlPart, currencyCode},
        },
        lazyload: {colourchips},
        request: {isEnglishLang},
        text,
    } = state

    return {
        colourchips: (colourways as ColourwayData[]).map(colourway => ({
            itemNumber: colourway.id,
            colourChipUrl: colourway.colourChipImage,
            linkUrl: formatPdpLink(baseUrl, colourway.url),
            isSelected: colourway.selected,
            colour: colourway.colour,
            title: colourway.title,
            price: selectPriceType(colourway.price, colourway?.salePrice),
            currencyCode,
            department,
            altText: formatTitle({
                title,
                department,
                fit,
                selectedColourway: colourway,
                isMultipleColourways: colourways.length > 1,
                isEnglishLang,
            }),
            text,
        })),
        text,
        lazyloadColourchips: colourchips,
        colourWayImageUrlBuilder: (id: string) =>
            `${imageCdnUrl}/${productImageUrlPart}/${LG_IMAGE_PATH}/${id}${IMAGE_EXTENSION}`,
    }
}

export default connect(mapStateToProps)
