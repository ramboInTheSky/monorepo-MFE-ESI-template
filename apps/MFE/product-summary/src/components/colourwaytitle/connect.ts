import {connect} from "react-redux"
import {selectPriceType} from "../../utils/priceFormatter/selectPriceType"
import {ColourwayData} from "../../models/searchApi"
import State from "../../models/state"
import {formatTooltipTitle, formatPdpLink} from "../../utils/colourwayBuilders"
import getTitleUtil from "../../utils/getTitleUtil"

export const mapStateToProps = (state: State) => {
    const {
        productSummary: {
            summaryData: {id, title, colourways, baseUrl, fit, department, currencyCode},
            enabledSearchDesc,
        },
        request: {isEnglishLang},
    } = state

    const selectedColourway = (colourways as ColourwayData[]).find(colourway => colourway.selected)

    const getTitle = getTitleUtil({
        enabledSearchDesc,
        title,
        selectedColourway,
        fit,
        department,
        colourways,
        isEnglishLang,
    })

    return {
        id,
        title: getTitle,
        tooltipTitle: formatTooltipTitle(
            getTitle,
            selectedColourway!.id,
            selectedColourway!.price,
            selectedColourway!.salePrice,
        ),
        linkUrl: formatPdpLink(baseUrl, selectedColourway!.url),
        price: selectPriceType(selectedColourway!.price, selectedColourway?.salePrice),
        colour: selectedColourway!.colour,
        currencyCode,
        department,
    }
}

export default connect(mapStateToProps)
