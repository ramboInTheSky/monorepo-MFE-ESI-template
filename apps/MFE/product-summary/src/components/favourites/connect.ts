import {connect} from "react-redux"
import State from "../../models/state"
import {PublishRemoveFromFav, PublishAddToFav} from "../../events"
import {getListId} from "../../utils/getWindowNextFavourites"
import {setFavouritedColourways, setLoadingFav, setAnimateFavouriteIcon} from "../../ducks/productSummary"
import {PublishToModalsClosed} from "../../events/modalsClosed"

export const mapStateToProps = ({productSummary, text}: State) => {
    const {
        isFav,
        showFavErrorToolTip,
        enabledSearchDesc,
        isLoadingFav,
        animateFavouriteIcon,
        summaryData: {id, title, baseUrl, colourways, fit, department},
    } = productSummary
    return {
        isFav,
        itemNumber: id,
        title,
        showFavErrorToolTip,
        baseUrl,
        colourways,
        enabledSearchDesc,
        fit,
        department,
        isLoadingFav,
        text,
        animateFavouriteIcon,
        selectedColourwayTitle: colourways.find(colourway => colourway.selected)?.title,
    }
}

export const mergeProps = (state: any, {dispatch}: any, ownProps) => ({
    ...state,
    ...ownProps,
    addToFavourites: () => {
        dispatch(setLoadingFav(true))
        dispatch(setAnimateFavouriteIcon(true))
        PublishToModalsClosed()
        const eventId = Date.now().toString()
        const listId = getListId()
        const data = {
            itemNumber: state.colourways.find(colourway => colourway.selected)?.id,
            listId,
            optionCode: -1,
            eventId,
        }
        dispatch(PublishAddToFav(data))
    },

    removeFromFavourites: () => {
        dispatch(setLoadingFav(true))
        dispatch(setAnimateFavouriteIcon(true))
        PublishToModalsClosed()
        const eventId = Date.now().toString()
        const listId = getListId()
        const data = {
            itemNumber: state.colourways.find(colourway => colourway.selected)?.id,
            listId,
            optionCode: -1,
            eventId,
        }
        dispatch(PublishRemoveFromFav(data))
    },
    setFavouritedColourways: (favouritedColourways: string[]) =>
        dispatch(setFavouritedColourways(favouritedColourways)),
})

export default connect(mapStateToProps, null as any, mergeProps)
