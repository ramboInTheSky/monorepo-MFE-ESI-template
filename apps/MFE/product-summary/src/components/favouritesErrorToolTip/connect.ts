import {connect} from "react-redux"
import State from "../../models/state"
import {setShowFavErrorToolTip} from "../../ducks/productSummary"
import {FavErrorToolTipType} from "../../models/ProductSummary"

export const mapStateToProps = ({productSummary, text}: State) => ({
    showFavErrorToolTip: productSummary.showFavErrorToolTip,
    baseUrl: productSummary.summaryData.baseUrl,
    text
})

export const mapDispatchToProps = (dispatch: any) => ({
    setShowFavErrorToolTip: (show: FavErrorToolTipType | null) => dispatch(setShowFavErrorToolTip(show)),
})

export default connect(mapStateToProps, mapDispatchToProps)
