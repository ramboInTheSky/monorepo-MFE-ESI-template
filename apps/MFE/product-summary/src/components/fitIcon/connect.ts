import {connect} from "react-redux"

import State from "../../models/state"

export const mapStateToProps = (state: State) => ({
    isLazyloadFitIcons: state.lazyload.fitIcons,
    text: state.text
})

export default connect(mapStateToProps)