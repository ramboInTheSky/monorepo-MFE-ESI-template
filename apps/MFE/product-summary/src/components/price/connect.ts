import {connect} from "react-redux"
import State from "../../models/state"
import {formatTooltipTitle, formatPdpLink} from "../../utils/colourwayBuilders"
import getTitleUtil from "../../utils/getTitleUtil"

export const mapStateToProps = (state: State) => {
    const {
        productSummary: {
            summaryData: {title, colourways, baseUrl, id, fit, department, currencyCode, brand},
            enabledSearchDesc,
            enableBrandDisplay,
        },
        request: {isEnglishLang},
        text,
    } = state

    const selectedColourway = colourways.find(colourway => colourway.selected)
    const colourwaysHasSale = colourways.some(colourway => colourway.salePrice !== null)
    const colourwayWithPersonalisedType = selectedColourway!.personalisedType

    const getTitle = getTitleUtil({
        enabledSearchDesc,
        title,
        selectedColourway,
        fit,
        department,
        colourways,
        isEnglishLang,
    })

    const brandNameEnabled = enableBrandDisplay && brand !== null
    return {
        price: selectedColourway!.price,
        tooltipTitle: formatTooltipTitle(
            getTitle,
            selectedColourway!.id,
            selectedColourway!.price,
            selectedColourway!.salePrice,
        ),
        linkUrl: formatPdpLink(baseUrl, selectedColourway!.url),
        wasPrice: selectedColourway!.wasPrice,
        salePrice: selectedColourway!.salePrice,
        id,
        title: getTitle,
        colour: selectedColourway!.colour,
        currencyCode,
        department,
        text,
        colourwaysHasSale,
        brandNameEnabled,
        isMade2Measure: colourwayWithPersonalisedType === "madeToMeasure",
        minPrice: selectedColourway!.minPrice
    }
}

export default connect(mapStateToProps)
