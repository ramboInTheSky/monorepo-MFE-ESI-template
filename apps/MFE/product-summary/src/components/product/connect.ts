import {connect} from "react-redux"
import State from "../../models/state"
import setFavouriteEnabled from "../../utils/setFavouriteEnabled"
import isMadeToMeasure from "../../utils/isMadeToMeasure"
import {ProductType} from "../../config/constants"

export const mapStateToProps = (state: State) => {
    const {
        productSummary: {
            summaryData: {id: itemNumber, type: productType, colourways, brand},
            enableFav,
            enableReviewStars,
            enableBrandDisplay,
        },
    } = state
    const selectedColourway = colourways.find(colourway => colourway.selected)
    const isMade2Measure = isMadeToMeasure(selectedColourway!.personalisedType)
    const isPersonalised = !colourways[0].personalisedType
    return {
        itemNumber,
        favouriteBtnEnabled: (enableFav && !isMade2Measure) ? setFavouriteEnabled(productType, isPersonalised) : false,
        productType,
        enableReviewStars: enableReviewStars ? productType === ProductType : false,
        enableBrandDisplay,
        brand
    }
}

export default connect(mapStateToProps)
