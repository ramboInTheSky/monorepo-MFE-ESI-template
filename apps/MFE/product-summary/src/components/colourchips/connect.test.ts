import {mockState, mockStateSofa} from "../../../__mocks__/mockStore"
import {mapStateToProps} from "./connect"
import {selectPriceType} from "../../utils/priceFormatter/selectPriceType"
import text from "../../../__mocks__/default-text.json"

jest.mock("../../utils/getTitleUtil", () => ({
    __esModule: true,
    default: jest.fn(() => "colour pdp title"),
}))

describe("Given connect - mapStateToProps()", () => {
    const mappedProps = mapStateToProps(mockState)

    it("should project state and only return pid", () => {
        const altTextArr = ["White tall My Product", "Black tall My Product"]
        expect(mappedProps).toEqual({
            colourchips: (mockState.productSummary.summaryData as any).colourways.map((colourway, idx) => ({
                itemNumber: colourway.id,
                linkUrl: `${mockState.productSummary.summaryData.baseUrl}/${colourway.url}`,
                isSelected: colourway.selected,
                colour: colourway.colour,
                colourChipUrl: `${mockState.productSummary.summaryData.imageCdnUrl}/AltItemSwatch/21x21/${colourway.id}.jpg`,
                currencyCode: mockState.productSummary.summaryData.currencyCode,
                price: selectPriceType(colourway.price, colourway?.salePrice),
                title: colourway.title,
                altText: altTextArr[idx],
                text,
                department: "menswear",
            })),
            lazyloadColourchips: false,
            colourWayImageUrlBuilder: expect.any(Function),
            text,
        })
    })

    describe("when product is a sofa", () => {
        const res = mapStateToProps(mockStateSofa)
        it("should return expected data", () => {
            expect(res).toEqual({
                colourchips: (mockStateSofa.productSummary.summaryData as any).colourways.map(colourway => ({
                    itemNumber: colourway.id,
                    linkUrl: `${mockStateSofa.productSummary.summaryData.baseUrl}/${colourway.url}`,
                    isSelected: colourway.selected,
                    colour: colourway.colour,
                    colourChipUrl: colourway.colourChipImage,
                    currencyCode: mockStateSofa.productSummary.summaryData.currencyCode,
                    price: selectPriceType(colourway.price, colourway?.salePrice),
                    title: colourway.title,
                    altText: "Jackson",
                    text,
                    department: "homeware",
                })),
                lazyloadColourchips: false,
                colourWayImageUrlBuilder: expect.any(Function),
                text,
            })
        })
    })

    it("should render the right url the 'colourWayImageUrlBuilder'", () => {
        expect(mappedProps.colourWayImageUrlBuilder("my-item-number")).toMatchSnapshot()
    })
})
