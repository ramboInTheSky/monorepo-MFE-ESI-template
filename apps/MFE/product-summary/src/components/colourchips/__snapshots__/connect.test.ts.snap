// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given connect - mapStateToProps() should render the right url the 'colourWayImageUrlBuilder' 1`] = `"http://testcdn.co.uk/search/224x336/my-item-number.jpg"`;
